fem/libceed/solvers-atpmg.cpp

```diff
@@ -0,0 +1,728 @@
+   }
+   else
+   {
+      coarse_i = -1;
+   }
+   if (i == 0)
+   {
+      coarse_i = 0;
+   }
+   return coarse_i;
+}
+
+int min4(int a, int b, int c, int d)
+{
+   if (a <= b && a <= c && a <= d)
+   {
+      return a;
+   }
+   else if (b <= a && b <= c && b <= d)
+   {
+      return b;
+   }
+   else if (c <= a && c <= b && c <= d)
+   {
+      return c;
+   }
+   else
+   {
+      return d;
+   }
+}
+
+int CeedATPMGElemRestriction(int order,
+                             int order_reduction,
+                             CeedElemRestriction er_in,
+                             CeedElemRestriction* er_out,
+                             CeedInt *&dof_map)
+{
+   int ierr;
+   Ceed ceed;
+   ierr = CeedElemRestrictionGetCeed(er_in, &ceed); CeedChk(ierr);
+
+   CeedInt numelem, numnodes, numcomp, elemsize;
+   ierr = CeedElemRestrictionGetNumElements(er_in, &numelem); CeedChk(ierr);
+   ierr = CeedElemRestrictionGetLVectorSize(er_in, &numnodes); CeedChk(ierr);
+   ierr = CeedElemRestrictionGetElementSize(er_in, &elemsize); CeedChk(ierr);
+   ierr = CeedElemRestrictionGetNumComponents(er_in, &numcomp); CeedChk(ierr);
+   if (numcomp != 1)
+   {
+      // todo: this will require more thought
+      return CeedError(ceed, 1, "Not implemented!");
+   }
+
+   int P1d = order + 1;
+   int coarse_P1d = P1d - order_reduction;
+   int dim = (log((double) elemsize) / log((double) P1d)) + 1.e-3;
+
+   CeedVector in_lvec, in_evec;
+   ierr = CeedElemRestrictionCreateVector(er_in, &in_lvec, &in_evec);
+   CeedChk(ierr);
+
+   CeedScalar * lvec_data;
+   ierr = CeedVectorGetArray(in_lvec, CEED_MEM_HOST, &lvec_data); CeedChk(ierr);
+   for (int i = 0; i < numnodes; ++i)
+   {
+      lvec_data[i] = (CeedScalar) i;
+   }
+   ierr = CeedVectorRestoreArray(in_lvec, &lvec_data); CeedChk(ierr);
+
+   // todo: I am making assumptions about the ordering of the evec that in
+   // principle are decided by the backend, which I do not control
+   ierr = CeedElemRestrictionApply(er_in, CEED_NOTRANSPOSE, in_lvec, in_evec,
+                                   CEED_REQUEST_IMMEDIATE); CeedChk(ierr);
+   ierr = CeedVectorDestroy(&in_lvec); CeedChk(ierr);
+   const CeedScalar * in_elem_dof;
+   ierr = CeedVectorGetArrayRead(in_evec, CEED_MEM_HOST, &in_elem_dof);
+   CeedChk(ierr);
+
+   // map high-order ldof to low-order ldof
+   // !! caller's responsibility to free
+   dof_map = (CeedInt*) calloc(numnodes, sizeof(CeedInt));
+   for (int i = 0; i < numnodes; ++i)
+   {
+      dof_map[i] = -1;
+   }
+
+   const bool debug = false;
+   CeedInt coarse_elemsize = pow(coarse_P1d, dim);
+   if (debug)
+      printf("coarse_P1d = %d, numelem = %d, coarse_elemsize = %d\n",
+             coarse_P1d, numelem, coarse_elemsize);
+   CeedInt * out_elem_dof = (CeedInt*) calloc(coarse_elemsize * numelem,
+                                              sizeof(CeedInt));
+
+   int running_out_ldof_count = 0;
+   if (dim == 2)
+   {
+      for (int e = 0; e < numelem; ++e)
```

Comment:
Do you have any idea if this loop could use `MFEM_FORALL`?

Reply:
`MFEM_FORALL` is not about parallelism, it's about using a device.

Reply:
To clarify, threading (on-node parallelism) will not work here because each iteration of the loop has to run sequentially, or else the produced numbering will not be correct.